pkg/util/clusterdata/worker_profiles_task.go

```diff
@@ -0,0 +1,113 @@
+package clusterdata
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	azureproviderv1beta1 "github.com/openshift/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
+	clusterapi "github.com/openshift/cluster-api/pkg/client/clientset_generated/clientset"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+const (
+	workerMachineSetsNamespace = "openshift-machine-api"
+)
+
+func newWorkerProfilesEnricherTask(log *logrus.Entry, restConfig *rest.Config, oc *api.OpenShiftCluster) (enricherTask, error) {
+	client, err := clusterapi.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &workerProfilesEnricherTask{
+		log:    log,
+		client: client,
+		oc:     oc,
+	}, nil
+}
+
+type workerProfilesEnricherTask struct {
+	log    *logrus.Entry
+	client clusterapi.Interface
+	oc     *api.OpenShiftCluster
+}
+
+func (ef *workerProfilesEnricherTask) FetchData(callbacks chan<- func()) {
+	r, err := azure.ParseResourceID(ef.oc.ID)
+	if err != nil {
+		ef.log.Error(err)
+		callbacks <- func() {}
+		return
+	}
+
+	machinesets, err := ef.client.MachineV1beta1().MachineSets(workerMachineSetsNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		ef.log.Error(err)
+		callbacks <- func() {}
+		return
+	}
+
+	if len(machinesets.Items) == 0 {
+		ef.log.Errorf("worker machinesets not found in %q", workerMachineSetsNamespace)
+		callbacks <- func() {}
+		return
+	}
+
+	workerProfiles := make([]api.WorkerProfile, len(machinesets.Items), len(machinesets.Items))
+	for i, machineset := range machinesets.Items {
+		var machineProviderSpec azureproviderv1beta1.AzureMachineProviderSpec
+		_, _, err := codecs.UniversalDeserializer().Decode(machineset.Spec.Template.Spec.ProviderSpec.Value.Raw, nil, &machineProviderSpec)
```

Comment:
unfortunately this will not detect if you are passed an unexpected groupversionkind :-(

do:
```
o, _, err := codecs.UniversalDeserializer().Decode(ms.Spec.Template.Spec.ProviderSpec.Value.Raw, nil, nil)
if err != nil {
	// ...
}
if o, ok := o.(*azureproviderv1beta1.AzureMachineProviderSpec); ok { // or whatever
	// ...
}
```

Reply:
note: because of what I suspect is a bug caused by forking [here](https://github.com/openshift/cluster-api-provider-azure/blob/master/pkg/apis/addtoscheme_azureprovider_v1beta1.go#L20), for the above to work you will need to do
```
	err = azureproviderv1beta1.SchemeBuilder.AddToScheme(scheme)
	if err != nil {
		return err
	}
```
instead of
```
	utilruntime.Must(azureproviderapis.AddToScheme(scheme))
```